client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+The cache entries themselves are stored in these arraylists.  The values are tuples derived from the value timestamps and `value : read-time` concatenated strings.  
+
+If the arraylists are unsorted, then `add` is *O(1)*, and `find` (floor), `replace`, and `remove` are each *O(n)*, where `n` is the length of the arraylist.  If the arraylists are sorted (which takes *O(nlogn)*), all the operations are *O(n)*.  Thus, the arraylists do not need to be sorted, as sorting would not improve the asymptotic runtimes of these operations.
+
+Note that the cache can indicate the DB did not have a value for a certain key at a given time by storing None in Python (null in other languages) in the `(value, read-time)` entry. 
+
+#### Skiplist Implementation
+
+If an appropriate skiplist library is available, the cache can be implemented as a skiplist which supports `floor` and `ceiling` operations.  The cache is then a single-level map and has the following type:
+
+    table : key : (Long.Max - ValueTime) -> (value, read-time)
+
+Here is an example of a [skiplist](https://github.com/sepeth/python-skiplist) library in Python, though it does not support floor and ceiling operations.
+
+## Cache Public Method Descriptions
+
+### read
+    read(
+        table: string, 
+        key: string,
+        max_age=None, 
+        no_cache=False): value (JSON)  
+
+The read method first checks the cache.  If the entry is missing, or if it has been too long since we read that entry, it sends a read request to the DB for the given table and key, along with the specified cache parameters, if any.  It returns the value object if the DB provides a value; otherwise, it returns None.
+
+In the case read returns None (value not in database), we still update our cache with the value entry None for the specified key using the read\_timestamp.  The None value may represent an update to the key.
+
+Note that the max\_age, and no\_cache parameters can be specified by the user in at least three separate locations: (1) in the Cache constructor, (2) in the Cache read method, and (3) in the Transaction read method.  The cache read method determines the most restrictive set of parameters based on these three inputs and sends this most restrictive parameter set to the DB.
+
+Here is how we define the most restrictive parameter set:
+
+* **max\_age**: The most restrictive max\_age is the minimum of all given max\_age values (which we formalize [here](#minRt)).
+
+* **no\_cache**: The most restrictive no\_cache value is the disjunction (boolean "or") of all given no\_cache values.  (If no\_cache == True, the cache must verify with the server by checking the value timestamp it has the most up-to-date value.)
+
+Note that these max\_age and no\_cache values also appear in Cache read method, though users cannot control their values through this method.
+
+### write
+    write(
+        table: string, 
+        op_List: list): value (JSON) throws StaleException  
```

Comment:
The ops are the map from the Tx view, not a list. This text should probably point to the discussion of Tx.commit, or vice versa.